lib/DocumentService.php

```diff
@@ -142,17 +157,62 @@ public function getDocumentByUserId(string $userId, int $fileId, ?string $dir) :
 				return $this->reportError('Document for the fileId ' . $fileId . 'not found');
 			}
 
-			// Set basic parameters
+			/** @var \OCP\Files\Storage\IStorage $storage */
+			$storage = $document->getStorage();
+			$isSharedFile = $storage->instanceOfStorage('\OCA\Files_Sharing\SharedStorage');
+			$isFederatedShare = $storage->instanceOfStorage('\OCA\Files_Sharing\External\Storage');
+			$isSecureModeEnabled = $this->appConfig->secureViewOptionEnabled();
+
+			// Base file info
+			$ret = [];
 			$ret['owner'] = $document->getOwner()->getUID();
-			$ret['permissions'] = $document->getPermissions();
-			$ret['updateable'] = $document->isUpdateable();
+			$ret['allowEdit'] = $document->isUpdateable();
+			$ret['allowExport'] = true;
+			$ret['allowPrint'] = true;
+			$ret['secureView'] = false;
+			$ret['secureViewId'] = null;
+			$ret['federatedServer'] = null;
+			$ret['federatedToken'] = null;
 			$ret['mimetype'] = $document->getMimeType();
 			$ret['path'] = $root->getRelativePath($document->getPath());
 			$ret['name'] = $document->getName();
 			$ret['fileid'] = $fileId;
-			$ret['instanceid'] = $this->config->getSystemValue('instanceid');
 			$ret['version'] = '0'; // latest
 
+			if ($isSharedFile && $isSecureModeEnabled) {
+				/** @var \OCA\Files_Sharing\SharedStorage $storage */
+				/* @phan-suppress-next-line PhanUndeclaredMethod */
+				$share = $storage->getShare();
+				$sharePermissionsDownload = $share->getAttributes()->getAttribute('permissions', 'download');
+				$shareViewWithWatermark = $share->getAttributes()->getAttribute('richdocuments', 'view-with-watermark');
+				$shareCanPrint = $share->getAttributes()->getAttribute('richdocuments', 'print');
+
+				// restriction on view has been set to false, return forbidden
+				// as there is no supported way of opening this document
+				if ($sharePermissionsDownload === false && $shareViewWithWatermark === false) {
+					return $this->reportError('Insufficient file permissions for the fileId ' . $fileId);
+				}
+
+				// can export file in editor if download is not set or true
+				$ret['allowExport'] = ($sharePermissionsDownload === null || $sharePermissionsDownload === true);
+
+				// can print from editor if print is not set or true
+				$ret['allowPrint'] = ($shareCanPrint === null || $shareCanPrint === true);
+
+				// view with watermarking enabled with private mode enabled
+				if ($shareViewWithWatermark === true) {
+					$ret['allowEdit'] = false;
+					$ret['secureView'] = true;
+					$ret['secureViewId'] = $share->getId();
+				}
+			} elseif ($isFederatedShare) {
+				/** @var \OCA\Files_Sharing\External\Storage $storage */
+				/* @phan-suppress-next-line PhanUndeclaredMethod */
+				$ret['federatedServer'] = $storage->getRemote();
+				/* @phan-suppress-next-line PhanUndeclaredMethod */
+				$ret['federatedToken'] = $storage->getToken();
+			}
```

Comment:
moved all code related to shares and permissions/attributes into DocumentService. Now DocumentController is not aware of this business logic related to UserFolder, external/share storages etc. It simplifies and isolates that complexity.